src/ai/backend/agent/config/unified.py

```diff
@@ -1314,3 +1396,36 @@ def _validate_agent_configs(self) -> Self:
             config.validate_agent_specific_config()
 
         return self
+
+    @model_validator(mode="after")
+    def _validate_resource_allocation_mode(self) -> Self:
+        agent_configs = self.get_agent_configs()
+
+        match self.resource.allocation_mode:
+            case ResourceAllocationMode.SHARED | ResourceAllocationMode.AUTO_SPLIT:
+                for config in agent_configs:
+                    resource = config.resource
+                    if any([
+                        resource.allocated_cpu is not None,
+                        resource.allocated_mem is not None,
+                        resource.allocated_devices,
+                    ]):
+                        raise ValueError(
+                            "On non-MANUAL mode, config must not specify manual resource allocations"
+                        )
+
+            case ResourceAllocationMode.MANUAL:
+                for config in agent_configs:
+                    resource = config.resource
+                    if any([resource.allocated_cpu is None, resource.allocated_mem is None]):
+                        raise ValueError(
+                            "On MANUAL mode, config must specify cpu and mem resource allocations"
+                        )
```

Comment:
It seems more appropriate to specify this at the type level. Allowing a type to be optional when it shouldn’t be, and then validating it at runtime, feels inconsistent.